### DISCUSSION TITLE
setStencilReference for GPURenderBundleEncoder

### DISCUSSION THREAD

**Discussion by Pauan** [gpuweb/gpuweb#4278] [Q&A] [answered]
Currently there is a [`GPURenderPassEncoder.setStencilReference`](https://developer.mozilla.org/en-US/docs/Web/API/GPURenderPassEncoder/setStencilReference) method, but it is missing from [`GPURenderBundleEncoder`](https://developer.mozilla.org/en-US/docs/Web/API/GPURenderBundleEncoder).

Is there a reason for this omission? It would be useful for my use case (since my `GPURenderBundle` needs to set the stencil reference value).

**Comment by Pauan** [comment]
As a workaround, it's possible to interleave `setStencilReference` with `executeBundles`, like this:

```js
renderPass.setStencilReference(0);
renderPass.executeBundles([bundle1]);

renderPass.setStencilReference(1);
renderPass.executeBundles([bundle2]);

renderPass.setStencilReference(2);
renderPass.executeBundles([bundle3]);
```

Is this behavior guaranteed by the spec? What I mean is:

* When `bundle1` is executed, is it guaranteed to have a stencil reference of `0`?
* When `bundle2` is executed, is it guaranteed to have a stencil reference of `1`?
* When `bundle3` is executed, is it guaranteed to have a stencil reference of `2`?

**Reply by kainino0x** [reply]
Here's the relevant investigation:
https://github.com/gpuweb/gpuweb/issues/286#issuecomment-492865792
But I need to think more about your question, so I will get back to you.

**Comment by kainino0x** [comment, marked as answer]
WebGPU render bundles are meant to abstract over three native API concepts: Vulkan reusable secondary command buffers, D3D12 bundles, and Metal indirect command buffers. According to the investigation (https://github.com/gpuweb/gpuweb/issues/286#issuecomment-492865792), Metal is the reason that setStencilReference is not available in render bundles.

In WebGPU, there are two "tiers" of render state: that which can only be set in a render pass (and thus must be inherited into render bundles), and that which can _also_ be set in a render bundle. WebGPU is strict about the inheritance here, so the latter tier is **not** inherited into render bundles, and the pass's state is always cleared out when you call executeBundles.

On the Vulkan and D3D12 backends, IIRC we rely on these states being both settable inside the backend equivalents, *and* inheritable from the surrounding context. Thus allowing us to create a single abstraction that works consistently across backends. (This was the part I had to think more about.)

To answer your question - it is guaranteed by the spec. Though it could be a bit more explicit about it. It's implicit in the fact that those states are not reset by [executeBundles()](https://gpuweb.github.io/gpuweb/#dom-gpurenderpassencoder-executebundles), and that these states are on the [RenderState](https://gpuweb.github.io/gpuweb/#renderstate) rather than the [GPURenderCommandsMixin](https://gpuweb.github.io/gpuweb/#gpurendercommandsmixin) (shared by GPURenderPassEncoder and GPURenderBundleEncoder). (aside: there's some silly reason why RenderState is on the GPUCommandBuffer instead of the GPURenderPassEncoder, you can ignore that, it might get refactored eventually)

**Reply by Pauan** [reply]
> Metal is the reason that setStencilReference is not available in render bundles.

That makes sense. If the underlying API doesn't support it then there isn't much WebGPU can do about it. So I'll just need to use the interleaving workaround instead.

> To answer your question - it is guaranteed by the spec.

Thanks! Based on my reading of the spec [I believed that it was guaranteed](https://github.com/gfx-rs/wgpu/issues/4104#issuecomment-1702214859), because the spec says that the RenderState is inherited, but it's good to have it confirmed.
